src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp

```diff
@@ -790,7 +793,9 @@ class ShenandoahGenerationalUpdateHeapRefsTask : public WorkerTask {
 
     ShenandoahHeapRegion* r = _regions->next();
     // We update references for global, old, and young collections.
-    assert(_heap->active_generation()->is_mark_complete(), "Expected complete marking");
+    ShenandoahGeneration* const gc_generation = _heap->gc_generation();
```

Comment:
It would certainly involve more changes, but we could pass the correct generation from `shConcurrentGC` in this case. I think we could do this in most cases (save for the barriers, but you've established a safe pattern for them to access the  active generation already).

Reply:
Let'd defer that cleanup for later.